data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,444 @@
+---
+id: metaprogramming
+title: Preprocessors and PPXes
+description: >
+  An introduction to metaprogramming in OCaml, including preprocessors, ppx extensions and the ppxlib library.
+category: "guides"
+date: 2022-07-25T21:07:30-00:00
+---
+
+
+# Preprocessors and PPXes
+
+Preprocessors are programs meant to be called at compile-time, so that they
+alter the program's source code before the actual compilation. They can be very
+useful for several things, such as the inclusion of a file, conditional
+compilation, or use of macros.
+
+To start with an example, taking inspiration from the syntax from the C
+preprocessor, we could write something like that:
+
+
+```ocaml
+let compiler_system = #SYSTEM
+
+let f =
+#if STDLIB > 2
+  Stdlib.new_function
+#else
+  Stdlib.old_function
+#endif
+```
+
+A preprocessor could, at compile time, replace the `#SYSTEM` macro by a string
+with the information of the system on which it has been compiled, and only
+include the line using the right function, which depends on the current version
+of the `Stdlib`.
+
+In OCaml, there are many preprocessors that are written by the community. OCaml
+supports the execution of two kinds of preprocessors: one that work on the
+source level, and the other that work on the [AST
+level](#ocamls-parsetree-the-first-ocamls-ast). The latter is called "PPX", an
+acronym for Pre-Processor eXtension.
+
+While both types of preprocessing have their use-cases, in OCaml it is
```

Comment:
It's weird to read about all the great advantages and then suddenly talk about something else.

I suggest moving the source processing to after the PPX section. and don't start with "PPxes are a different kind of preprocessors".

Reply:
I think speaking about source preprocessors before PPXes is important for people reading the guide who are not familiar with the subject.

- It's easier to understand what happens, no need for AST explanation
- It allows to point out what are the limitation of this simple (and better known) approach, and so _why_ PPXes were introduced, which makes them easier to understand.

I added a line explaining why we keep this order in the guide.

Reply:
Sounds good.